README.md

```diff
@@ -215,3 +215,55 @@ Currently, there are three policy configurations.
 
 A build Task, e.g. one that produces a container image, must abide by both `all-tasks` and
 `build-tasks` policy configurations.
+
+## Task Migration
+
+Task migrations are Bash scripts defined in version-specific task
+directories. In general, a migration consists of a series of `yq` commands that
+modify pipeline in order to work with the new version of task.
+
+Historically, task maintainers write `MIGRATION.md` to notify users what changes
+have to be made to the pipeline. This mechanism is not deprecated. Besides
+writing the file, it is also recommended to write a migration so that the
+updates can be applied to user pipelines automatically.
+
+The following is the steps to write a migration:
+
+- Bump task version. Modify label `app.kubernetes.io/version` in the task YAML file.
+- Ensure `migrations/` directory exists in the version-specific task directory.
+- Create a migration file under the `migrations/` directory. Its name is in
+  form `<new task version>.sh`. Note that the version must match the bumped
+  version.
+
+The migration file is a normal Bash script file. It should
+
+- Accept one argument. The pipeline file path is passed to via this argument.
+- All modifications to the pipeline must be done in-place, i.e. using `yq
+  -i` to operate the pipeline YAML.
+- Be idempotent as much as possible.
+- Be simple and small as much as possible.
+- Pass the `shellcheck` without customizing the default rules.
+
+Here are example script to create a migration for a task `task-a`:
+
+```bash
+yq -i "(.metadata.labels.\"app.kubernetes.io/version\") |= \"0.2.2\"" task/task-a/0.2/task-a.yaml
+mkdir -p task/task-a/0.2/migrations || :
+cat >task/task-a/0.2/migrations/0.2.2.sh <<EOF
+#!/usr/bin/env bash
+set -e
+pipeline_file=\$1
+
+if ! yq -e '.spec.tasks[] | select(.name == "task-a") | .params[] | select(.name == "pipelinerun-name")' >/dev/null
+then
+  yq -i -e '
+    (.spec.tasks[] | select(.name == "task-a") | .params) +=
+    {"name": "pipelinerun-name", "value": "\$(context.pipelineRun.name)"}
```

Comment:
Should we make this example idempotent? (I.e. only add the param if it's not already there)

Reply:
It's already idempotent. Line 257 checks that.

Reply:
Ah ok, I missed that. Maybe add a short comment about that above the `if`? It's not immediately obvious that that's why the check is there

Reply:
Added a comment for it.